File: ising.f90
program testeo
use mt19937

integer :: i,j,fila,columna,ia,ja,terma,mc,npasos
integer, allocatable:: s(:,:)
integer :: seed,trial_spin,contador,nf,nr,fm,nrm,jota=1
real :: w,r,energia,magne,magnesum,enersum,du,t,tf,ts,magnesum2,enersum2
real::fil,col,eq,np,simu
character(len=50):: archivo,nombre


call system_clock(count=seed)
call sgrnd(seed)

read*,fil
read*,col
read*,t
read*,tf
read*,ts
read*,eq
read*,np
read*, simu

fila=int(fil)
columna=int(col)
terma=int(eq)
npasos=int(np)

print*,"Input Parameters"
print*,"---------------------------------------"
print*,"Rows: ", fila
print*,"Columns: ", columna
print*,"Initial Temperature: ",t
print*,"Final Temperature: ",tf
print*,"Temperature Step: ",ts
print*,"Iterations until Termalization: ",terma
print*,"MonteCarlo Iterations: ",npasos
print*,"---------------------------------------"

Allocate(s(fila,columna))
open(97,file="results.txt",status="unknown")
open(99,file="spin_f.txt",status="unknown")
open(98,file="spin.txt",status="unknown")

do i=1,fila
	read(98,*) s(i,:)
end do
close(98)



do while(t.le.tf)




contador=0
du=0
enersum=0
magnesum=0
energia=0
magne=0
magnesum2=0
enersum2=0
do mc=0,npasos


	if(mc>terma) then
		contador=contador+1
		magne=sum(s(1:fila,1:columna))/(fila*columna*1.0)
		magnesum=magnesum+magne
		magnesum2=magnesum2+magne**2

		

		do i=1,fila
			do j=1,columna
				nf=0
				nr=0
				nfm=0
				nrm=0
				if(i==1) then
					nf=i-1+fila
					end if
				if (j==1) then
					nr=j-1+columna
				end if
				if (i==fila) then
					nfm=-i
				end if
				if (j==columna) then
					nrm=-j
				end if
				energia=energia-jota*s(i,j)*(s(i-1+nf,j)+s(i+1+nfm,j)+s(i,j-1+nr)+s(i,j+1+nrm))
				
			end do
		end do
		energia=energia/(fila*columna*2)
		enersum=enersum+energia
		enersum2=enersum2+energia**2

	end if
	ia=aleatorio(1,fila)
	ja=aleatorio(1,columna)
	trial_spin=-s(ia,ja)
	nf=0
	nr=0
	nfm=0
	nrm=0
	if(ia==1) then
		nf=ia-1+fila
		end if
	if (ja==1) then
		nr=ja-1+columna
	end if
	if (ia==fila) then
		nfm=-ia
	end if
	if (ja==columna) then
		nrm=-ja
	end if

	du=-2*jota*trial_spin*(s(ia+1+nfm,ja)+s(ia,ja+1+nrm)+s(ia-1+nf,ja)+s(ia,ja-1+nr))
	if (du.le.0) then
		s(ia,ja)=trial_spin
	end if
	if (du>0) then
		w=exp(-du/t)
		r=grnd()
		if (r.le.w) then
			s(ia,ja)=trial_spin
		else
			s(ia,ja)=s(ia,ja)
		end if
	end if
	
	if (simu==1 .and. t==tf) then
		write(archivo,'(i0)') mc
		nombre="resultados\"//trim(adjustl(archivo))//".txt"
		open(96,file=trim(nombre),status="unknown")
		do i=1,fila
			write(96,*) s(i,:)
		end do
		close(96)
	end if
	
end do


write(97,*) enersum/contador,(t**-2)*(enersum2/contador-(enersum/contador)**2),&
abs(magnesum/contador),(1/t)*(magnesum2/contador-(magnesum/contador)**2),t
t=t+ts


end do

do i=1,fila
	write(99,*) s(i,:)
end do

contains

integer function aleatorio(a,b)
	integer :: a,b
	real :: r
	r=grnd()
	ja=nint((a+(b-a)*grnd()))
	return
end function

end program testeo

File: ising.py
import numpy as np
import random
import os as os
import matplotlib.pylab as plt


art=("_______________________________________________________________________________________________\n"
" _  _  _             _  _  _  _            _  _  _           _           _           _  _  _\n"       
"(_)(_)(_)          _(_)(_)(_)(_)_         (_)(_)(_)         (_) _       (_)       _ (_)(_)(_)(_)\n"   
"   (_)            (_)          (_)           (_)            (_)(_)_     (_)      (_)         (_)\n"   
"   (_)            (_)_  _  _  _              (_)            (_)  (_)_   (_)      (_)    _  _  _ \n"   
"   (_)              (_)(_)(_)(_)_            (_)            (_)    (_)_ (_)      (_)   (_)(_)(_)\n"   
"   (_)             _           (_)           (_)            (_)      (_)(_)      (_)         (_)\n"   
" _ (_) _          (_)_  _  _  _(_)         _ (_) _          (_)         (_)      (_) _  _  _ (_)\n"   
"(_)(_)(_)           (_)(_)(_)(_)          (_)(_)(_)         (_)         (_)         (_)(_)(_)(_)\n"
"_______________________________________________________________________________________________\n")    
print(art)

def aleatorio(A):
    for i in range(fil):
        for j in range(col):
            A[i,j]=random.choice(spin)

    np.savetxt("spin.txt",A,fmt='%1i')
    return A

def ajedrez(A):
    A[0,0]=1
    for i in range(fil-1):
        A[i+1,0]=-A[i,0]
    for j in range(col-1):
        A[:,j+1]=-A[:,j]
    np.savetxt("spin.txt",A,fmt='%1i')
    return A
        


def fundamental(A):
    for i in range(fil):
        for j in range(col):
            A[i,j]=1

    np.savetxt("spin.txt",A,fmt='%1i')
    return

fil=int(input("Number of Rows: "))
col=int(input("Number of Columns: "))
t=float(input("Initial Temperature: "))
tf=float(input("Final Temperature: "))
ts=float(input("Temperature Step: "))
equil=(input("Iterations until Termalization: "))
mc=int(input("MonteCarlo Iterations: "))
simu=int(input("Save Simulation 1-Yes, 2-No: "))
A=np.zeros([fil,col])
spin=[-1,1]

B=np.zeros([8,1])
B[0,0]=fil
B[1,0]=col
B[2,0]=t
B[3,0]=tf
B[4,0]=ts
B[5,0]=equil
B[6,0]=mc
B[7,0]=simu
np.savetxt("in.txt",B)
opcion={1:aleatorio,
        2:ajedrez,
        3:fundamental,
        }
        

print("Initial Configuration \n")
print("1-Ramdon\n2-Chess\n3-Fundamental")
seleccion=int(input())
A=opcion[seleccion](A)

opcion={1:"Ramdon",
        2:"Chess",
        3:"Fundamental",
        }

if not os.path.exists("results"):
    os.mkdir(f"{os.getcwd()}/results")
os.system("cls")
print(art)
print(f"Compiling using {fil}x{col} and spin {opcion[seleccion]}")
os.system("gfortran mt19937.f90 ising.f90 -o ising.exe")
print("Executing Program")
os.system(f"ising.exe <in.txt")
print("Done!!!!")
print("Plotting Initial & Final state")
fig,ax=plt.subplots(1,2)
Si=np.loadtxt("spin.txt")
Sf=np.loadtxt("spin_f.txt")

ax[0].imshow(Si)
ax[0].set_title("Initial State")
ax[1].imshow(Sf)
ax[1].set_title(f"Final State at T={tf}")
plt.show()




print("Plotting the Results!")
A=np.loadtxt("results.txt")
fig,ax =plt.subplots(2,2,figsize=(10,5))
plt.suptitle(f"{fil}x{col} Grid")
yaxis=["Energy","Cv","M","X"]
title=["Energy","Specific Heat","Magnetization","Susceptibility"]

k=0
for i in range(2):
    for j in range(2):
        ax[i,j].plot(A[:,4],A[:,k],"+")
        ax[i,j].set_title(f"{title[k]}")
        ax[i,j].set_xlabel("T")
        ax[i,j].set_ylabel(f"{yaxis[k]}")
        ax[i,j].grid()
        k=k+1
plt.show()

if (simu==1):
    j=0
    for i in os.listdir(f"{os.getcwd()}/results"):
        plt.cla()
        terma=np.loadtxt(f"{os.getcwd()}/results/{j}.txt")
        plt.title(f"iteracion {j}")
        plt.imshow(terma)
        plt.draw()
        plt.pause(0.01)
        j=j+1  